include/ofi_util.h

```diff
@@ -486,6 +486,75 @@ int ofi_get_src_addr(uint32_t addr_format,
 void ofi_getnodename(char *buf, int buflen);
 int ofi_av_get_index(struct util_av *av, const void *addr);
 
+/*
+ * Comamnd Queue
+ */
+
+struct util_cmd_handle {
+	struct dlist_entry entry;
+};
+
+typedef struct util_cmd_handle* (*util_cmd_queue_alloc)(size_t cmd_size);
+typedef void (*util_cmd_queue_free)(struct util_cmd_handle *handle);
+
+struct util_cmd_queue_attr {
+	util_cmd_queue_alloc cmd_alloc;
+	util_cmd_queue_free cmd_free;
+};
+
+struct util_cmd_queue {
+	struct dlist_entry queue;
+	struct util_cmd_queue_attr attr;
+	fastlock_t lock;
+};
+
+static inline
+int util_cmd_queue_empty(struct util_cmd_queue *queue)
+{
+	return dlist_empty(&queue->queue);
+}
+
+/* Retrieve a command from the head of CMD queue */
+static inline struct util_cmd_handle *
+util_cmd_queue_read(struct util_cmd_queue *queue)
+{
+	struct util_cmd_handle *handle;
+
+	if (util_cmd_queue_empty(queue))
```

Comment:
This check needs to be within the lock

Reply:
agreed, here we have to protect an access